add alternative solution (#6197)
cpinitiative/usaco-guide: solutions/silver/usaco-1063.mdx

</JavaSection>
</LanguageSection>

# Solution 2

## Explanation
Instead of keeping two prefix sums for the amount of cows above and below a y-coordinate, we can
simply keep one prefix sum. The amount of cows above and below a y-coordinate can then be calculated

> as a rectangle with the current y-coordinate and either extrema being the endpoints.

Reviewer: - should specify that the main difference is how we count cows above and below y coordinates
- would be clearer if you specified that we are just directly 2d prefix summing on coordinate compressed cow locations

## Implementation

**Time Complexity:** $\mathcal{O}(N^2)$

<LanguageSection>
<CPPSection>
```cpp
#include <bits/stdc++.h>

using namespace std;

bool sortByX(pair<int, int> a, pair<int, int> b) { return a.first < b.first; }

bool sortByY(pair<int, int> a, pair<int, int> b) { return a.second < b.second; }

long long rectSum(vector<vector<long long>> &grid, int r, int R, int c, int C) {
	return (grid[R][C] - grid[R][c - 1] - grid[r - 1][C] + grid[r - 1][c - 1]);

> }

Reviewer: - grid can be `const vector<vector<long long>> &grid`
- maybe r1, c1, r2, c2 as arguments? at least that's how I usually prefer to do it
- doesn't everything fit in an int here

int main() {
	int n;
	cin >> n;

	vector<vector<long long>> grid(n + 1, vector<long long>(n + 1));
	vector<pair<int, int>> nums(n);

> 	for (auto &e : nums) { cin >> e.first >> e.second; }

Reviewer: ```cpp
for (auto &[x, y] : nums) { cin >> x >> y; }
```

is cleaner IMO

	// compress X
	sort(nums.begin(), nums.end(), sortByX);
	for (int i = 0; i < nums.size(); i++) { nums[i].first = i + 1; }

	// compress Y
	sort(nums.begin(), nums.end(), sortByY);
	for (int i = 0; i < nums.size(); i++) { nums[i].second = i + 1; }

	// add points to grid and compute prefix
	for (auto [x, y] : nums) { grid[x][y] = 1; }

	for (int i = 1; i < grid.size(); i++) {
		for (int j = 1; j < grid[i].size(); j++) {
			grid[i][j] += grid[i - 1][j] + grid[i][j - 1] - grid[i - 1][j - 1];
		}
	}

	long long res = 0;
	// choose two points, count rectangles
	sort(nums.begin(), nums.end(), sortByX);
	for (int i = 0; i < n; i++) {
		for (int j = i + 1; j < n; j++) {
			int r = nums[i].first, R = nums[j].first;
			int c = min(nums[i].second, nums[j].second);
			int C = max(nums[i].second, nums[j].second);

			int above = rectSum(grid, 1, r - 1, c, C);
			int below = rectSum(grid, R + 1, n, c, C);

			res += (above + 1) * (below + 1);
		}
	}

> 	cout << res + (long long)n + 1 << "\n";

Reviewer: do we need to cast n as long long ?

also should add a comment for why adding n + 1

</JavaSection>
</LanguageSection>

# Solution 2

## Explanation
Instead of keeping two prefix sums for the number of cows above and below a y-coordinate, we can

> simply keep one prefix sum: a 2D prefix sum directly summing the coordinate compressed grid.

Reviewer 2: ```suggestion
keep only one 2D prefix sum over the coordinate-compressed grid.
```

The main difference between this and the solution above is that this solution uses rectangular
sums to find the points above and below a y-coordinate, while the solution above keeps track

> of the number of points above and below a y-coordinate.

Reviewer: ```suggestion
The central idea behind this solution is identical to the previous solution, but differs in how we count the cows above and below a bounding box.

Instead of keeping two prefix sums for the number of cows above and below a y-coordinate, we 
directly use 2D prefix sums on coordinate compressed locations. To count the number of cows above and below a bounding box, we use rectangular sums to directly sum the frequency of cows in the column above a box.
```

## Implementation

**Time Complexity:** $\mathcal{O}(N^2)$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

bool sortByX(pair<int, int> a, pair<int, int> b) { return a.first < b.first; }

> bool sortByY(pair<int, int> a, pair<int, int> b) { return a.second < b.second; }

Reviewer 2: 1. have these take const references
2. c++ uses snake case here

> int rectSum(vector<vector<long long>> &grid, int r1, int r2, int c1, int c2) {

Reviewer 2: ```suggestion
int rectSum(const vector<vector<long long>> &grid, int r1, int r2, int c1, int c2) {
```

int rectSum(vector<vector<long long>> &grid, int r1, int r2, int c1, int c2) {
	return (grid[r2][c2] - grid[r2][c1 - 1] - grid[r1 - 1][c2] + grid[r1 - 1][c1 - 1]);
}

int main() {
	int n;
	cin >> n;

	vector<vector<long long>> grid(n + 1, vector<long long>(n + 1));
	vector<pair<int, int>> nums(n);
	for (auto &[x, y] : nums) { cin >> x >> y; }

	// compress X
	sort(nums.begin(), nums.end(), sortByX);
	for (int i = 0; i < nums.size(); i++) { nums[i].first = i + 1; }

	// compress Y
	sort(nums.begin(), nums.end(), sortByY);
	for (int i = 0; i < nums.size(); i++) { nums[i].second = i + 1; }

	// add points to grid and compute prefix
	for (auto [x, y] : nums) { grid[x][y] = 1; }

	for (int i = 1; i < grid.size(); i++) {
		for (int j = 1; j < grid[i].size(); j++) {
			grid[i][j] += grid[i - 1][j] + grid[i][j - 1] - grid[i - 1][j - 1];
		}
	}

	long long res = 0;
	// choose two points, count rectangles
	sort(nums.begin(), nums.end(), sortByX);
	for (int i = 0; i < n; i++) {
		for (int j = i + 1; j < n; j++) {
			int r = nums[i].first, R = nums[j].first;
			int c = min(nums[i].second, nums[j].second);

> 			int C = max(nums[i].second, nums[j].second);

Reviewer: i think the variable naming scheme the prev solution uses was good (`r1`, `r2`, `bottom`, `top`)

also i think `R` should be on its own line

			int above = rectSum(grid, 1, r - 1, c, C);
			int below = rectSum(grid, R + 1, n, c, C);

			res += (above + 1) * (below + 1);
		}
	}

> 	// add empty rectangle

Reviewer: ```suggestion
	// add results for single or no cow
```
